testsuite/features/step_definitions/vm_steps.rb

```diff
@@ -0,0 +1,432 @@
+  raise 'not found: virsh' unless file_exists?(node, '/usr/bin/virsh')
+
+  node.run("virsh destroy #{vm_name}")
+end
+
+When(/^I delete the virtual machine named "([^"]*)" on "([^"]*)"$/) do |vm_name, host|
+  node = get_target(host)
+  raise 'not found: virsh' unless file_exists?(node, '/usr/bin/virsh')
+
+  step %(I stop the virtual machine named "#{vm_name}" on "#{host}")
+  node.run("virsh undefine --nvram #{vm_name}")
+end
+
+When(/^I create ([^ ]*) virtual network on "([^"]*)"$/) do |net_name, host|
+  node = get_target(host)
+
+  networks = {
+    'test-net0' => { 'bridge' => 'virbr0', 'subnet' => 124 },
+    'test-net1' => { 'bridge' => 'virbr1', 'subnet' => 126 },
+    'salt-sles' => { 'bridge' => 'br0' },
+    'salt-leap' => { 'bridge' => 'br0' },
+    'salt-rhlike' => { 'bridge' => 'br0' },
+    'salt-deblike' => { 'bridge' => 'br0' }
+  }
+
+  net = networks[net_name]
+  case net_name
+  when 'test-net0', 'test-net1'
+    netdef = '<network>' \
+              "  <name>#{net_name}</name>" \
+              "  <forward mode='nat'/>" \
+              "  <bridge name='#{net['bridge']}' stp='on' delay='0'/>" \
+              "  <ip address='192.168.#{net['subnet']}.1' netmask='255.255.255.0'>" \
+              "    <dhcp>" \
+              "      <range start='192.168.#{net['subnet']}.2' end='192.168.#{net['subnet']}.254'/>" \
+              "    </dhcp>" \
+              "  </ip>" \
+              "</network>"
+  when 'salt-sles', 'salt-leap', 'salt-rhlike', 'salt-deblike'
+    netdef = "<network>" \
+              "  <name>#{net_name}</name>" \
+              "  <forward mode='bridge'/>" \
+              "  <bridge name='#{net['bridge']}'/>" \
+              '</network>'
+  end
+
+  # Some networks like the default one may already be defined.
+  _output, code = node.run("virsh net-dumpxml #{net_name}", check_errors: false)
+  node.run("echo -e \"#{netdef}\" >/tmp/#{net_name}.xml && virsh net-define /tmp/#{net_name}.xml") unless code.zero?
+
+  # Ensure the network is started
+  node.run("virsh net-start #{net_name}", check_errors: false)
+end
+
+When(/^I delete ([^ ]*) virtual network on "([^"]*)"((?: without error control)?)$/) do |net_name, host, error_control|
+  node = get_target(host)
+  _output, code = node.run("virsh net-dumpxml #{net_name}", check_errors: false)
+  if code.zero?
+    steps %(
+      When I run "virsh net-destroy #{net_name}" on "#{host}"#{error_control}
+      And I run "virsh net-undefine #{net_name}" on "#{host}"#{error_control}
+    )
+  end
+end
+
+When(/^I create ([^ ]*) virtual storage pool on "([^"]*)"$/) do |pool_name, host|
+  node = get_target(host)
+
+  pool_def = %(<pool type='dir'>
+      <name>#{pool_name}</name>
+      <capacity unit='bytes'>0</capacity>
+      <allocation unit='bytes'>0</allocation>
+      <available unit='bytes'>0</available>
+      <source>
+      </source>
+      <target>
+        <path>/var/lib/libvirt/images/#{pool_name}</path>
+      </target>
+    </pool>
+  )
```

Comment:
Not saying we should refactor this, it's fine, but in other similar cases we extracted this inside upload_files, and then fill in.

Reply:
This is also something for another future PR. I just moved this step.